Firestore/core/test/firebase/firestore/remote/grpc_stream_test.cc

```diff
@@ -0,0 +1,318 @@
+  std::vector<std::string> States(std::initializer_list<std::string> states) {
+    return {states};
+  }
+
+  grpc::ClientContext* grpc_context = nullptr;
+  GrpcCompletionQueue grpc_queue;
+  std::unique_ptr<Observer> observer;
+  std::shared_ptr<GrpcStream> stream;
+};
+
+TEST_F(GrpcStreamTest, CannotStartTwice) {
+  EXPECT_NO_THROW(stream->Start());
+  EXPECT_ANY_THROW(stream->Start());
+}
+
+TEST_F(GrpcStreamTest, CannotWriteBeforeStreamIsOpen) {
+  EXPECT_ANY_THROW(stream->Write({}));
+  stream->Start();
+  EXPECT_ANY_THROW(stream->Write({}));
+}
+
+TEST_F(GrpcStreamTest, CanFinishBeforeStarting) {
+  EXPECT_NO_THROW(stream->Finish());
+}
+
+TEST_F(GrpcStreamTest, CanFinishAfterStarting) {
+  stream->Start();
+  EXPECT_NO_THROW(stream->Finish());
+}
+
+TEST_F(GrpcStreamTest, CannotFinishTwice) {
+  stream->Start();
+  EXPECT_NO_THROW(stream->Finish());
+  EXPECT_ANY_THROW(stream->Finish());
+}
+
+TEST_F(GrpcStreamTest, CanWriteAndFinishBeforeStarting) {
+  EXPECT_NO_THROW(stream->WriteAndFinish({}));
+}
+
+TEST_F(GrpcStreamTest, CanWriteAndFinishAfterStarting) {
+  stream->Start();
+  EXPECT_NO_THROW(stream->WriteAndFinish({}));
+}
+
+TEST_F(GrpcStreamTest, ObserverReceivesOnStart) {
+  stream->Start();
+  ForceFinish({/*Start*/ Ok});
+  EXPECT_EQ(observer->observed_states, States({"OnStreamStart"}));
+}
+
+TEST_F(GrpcStreamTest, CanWriteAfterStreamIsOpen) {
+  stream->Start();
+  ForceFinish({/*Start*/ Ok});
+  EXPECT_NO_THROW(stream->Write({}));
+}
+
+TEST_F(GrpcStreamTest, ObserverReceivesOnRead) {
+  stream->Start();
+  ForceFinish({/*Start*/ Ok, /*Read*/ Ok});
+  EXPECT_EQ(observer->observed_states,
+            States({"OnStreamStart", "OnStreamRead"}));
+}
+
+TEST_F(GrpcStreamTest, ReadIsAutomaticallyReadded) {
+  stream->Start();
+  ForceFinish({/*Start*/ Ok, /*Read*/ Ok});
+  EXPECT_EQ(observer->observed_states,
+            States({"OnStreamStart", "OnStreamRead"}));
+
+  ForceFinish({/*Read*/ Ok});
+  EXPECT_EQ(observer->observed_states,
+            States({"OnStreamStart", "OnStreamRead", "OnStreamRead"}));
+}
+
+TEST_F(GrpcStreamTest, ObserverReceivesOnWrite) {
+  stream->Start();
+  ForceFinish({/*Start*/ Ok});
+
+  stream->Write({});
+  ForceFinish({/*Read*/ Ignore, /*Write*/ Ok});
+
+  EXPECT_EQ(observer->observed_states,
+            States({"OnStreamStart", "OnStreamWrite"}));
+}
+
+TEST_F(GrpcStreamTest, CanAddSeveralWrites) {
+  stream->Start();
+  ForceFinish({/*Start*/ Ok});
+
+  stream->Write({});
+  stream->Write({});
+  stream->Write({});
+  ForceFinish({/*Read*/ Ignore, /*Write*/ Ok, /*Write*/ Ok, /*Write*/ Ok});
+
+  EXPECT_EQ(observer->observed_states,
+            States({"OnStreamStart", "OnStreamWrite", "OnStreamWrite",
+                    "OnStreamWrite"}));
+}
+
+TEST_F(GrpcStreamTest, ObserverReceivesOnError) {
+  stream->Start();
+  ForceFinish({/*Start*/ Ok});
+
+  // Fail the read, but allow Finish to succeed
+  ForceFinish({/*Read*/ Error, /*Finish*/ Ok});
```

Comment:
If we do not pass /*Finish*/ Ok, will the observer see the stream error?

Reply:
No longer applicable. Previously, no, the stream only notifies the observer of the error once `Finish` completes, because only then the stream will know the status. Finishing after an error is supposed to be very fast.